### DISCUSSION TITLE
cannot install 2.7.6 on mac 13.0 ventura m1

### DISCUSSION THREAD

**Discussion by aubri61** [rbenv/ruby-build#2125] [Build failures] [answered]
Hi, I was installing ruby 2.7.6 on rbenv but the error came out.
`CFLAGS="-Wno-error=implicit-function-declaration" RUBY_CONFIGURE_OPTS='--with-readline-dir=/usr/local/opt/readline/' arch -arm64 rbenv install 2.7.6`

> Inspect or clean up the working tree at /var/folders/j0/4ncpc9m90tg_c7d24lc1kvnr0000gn/T/ruby-build.20230113171746.72085.BaumO3
> Results logged to /var/folders/j0/4ncpc9m90tg_c7d24lc1kvnr0000gn/T/ruby-build.20230113171746.72085.log
> 
> Last 10 log lines:
> linking shared-object io/console.bundle
> compiling objspace_dump.c
> compiling date_strftime.c
> installing default syslog libraries
> installing digest libraries
> compiling date_strptime.c
> linking shared-object nkf.bundle
> linking shared-object objspace.bundle
> linking shared-object date_core.bundle
> make: *** [build-ext] Error 2

So I installed frum and tried this code

`RUBY_CFLAGS='-DUSE_FFI_CLOSURE_ALLOC -Wno-error=implicit-function-declaration'  frum install 2.7.6 --with-openssl-dir=$(brew --prefix openssl@1.0) --enable-shared`

And then this error came out

> Error: No available formula with the name "openssl@1.0". Did you mean openssl@1.1 or openssl@3?
==> Downloading https://cache.ruby-lang.org/pub/ruby/2.7/ruby-2.7.6.tar.xz
==> Extracting ruby-2.7.6.tar.xz
==> Building Ruby 2.7.6
error: Can't build Ruby: make install: 
Generating RI format into /Users/jueon/.frum/versions/.downloads/.tmp6dBID1/ruby-2.7.6/.ext/rdoc...
/Users/jueon/.frum/versions/.downloads/.tmp6dBID1/ruby-2.7.6/lib/rubygems/core_ext/kernel_require.rb:83:in `require': cannot load such file -- openssl (LoadError)
	from /Users/jueon/.frum/versions/.downloads/.tmp6dBID1/ruby-2.7.6/lib/rubygems/core_ext/kernel_require.rb:83:in `require'
	from /Users/jueon/.frum/versions/.downloads/.tmp6dBID1/ruby-2.7.6/lib/rubygems/specification.rb:2430:in `to_ruby'
	from ./tool/rbinstall.rb:846:in `block (2 levels) in install_default_gem'
	from ./tool/rbinstall.rb:279:in `open_for_install'
	from ./tool/rbinstall.rb:845:in `block in install_default_gem'
	from ./tool/rbinstall.rb:835:in `each'
	from ./tool/rbinstall.rb:835:in `install_default_gem'
	from ./tool/rbinstall.rb:799:in `block in <main>'
	from ./tool/rbinstall.rb:950:in `block in <main>'
	from ./tool/rbinstall.rb:947:in `each'
	from ./tool/rbinstall.rb:947:in `<main>'
make: *** [do-install-all] Error 1

So I tried to fix typo as openssol@1.1 but it said 

error: Version already installed at "/Users/jueon/.frum/versions/2.7.6"

so i deleted 2.7.6 dir and re installed it but the same error came out
`RUBY_CFLAGS='-DUSE_FFI_CLOSURE_ALLOC -Wno-error=implicit-function-declaration'  frum install 2.7.6 --with-openssl-dir=$(brew --prefix openssl@1.1) --enable-shared`

> ==> Downloading https://cache.ruby-lang.org/pub/ruby/2.7/ruby-2.7.6.tar.xz
> ==> Extracting ruby-2.7.6.tar.xz
> ==> Building Ruby 2.7.6
> error: Can't build Ruby: make install: 
> Generating RI format into /Users/jueon/.frum/versions/.downloads/.tmpm51eRo/ruby-2.7.6/.ext/rdoc...
> /Users/jueon/.frum/versions/.downloads/.tmpm51eRo/ruby-2.7.6/lib/rubygems/core_ext/kernel_require.rb:83:in `require': cannot load such file -- openssl (LoadError)
> 	from /Users/jueon/.frum/versions/.downloads/.tmpm51eRo/ruby-2.7.6/lib/rubygems/core_ext/kernel_require.rb:83:in `require'
> 	from /Users/jueon/.frum/versions/.downloads/.tmpm51eRo/ruby-2.7.6/lib/rubygems/specification.rb:2430:in `to_ruby'
> 	from ./tool/rbinstall.rb:846:in `block (2 levels) in install_default_gem'
> 	from ./tool/rbinstall.rb:279:in `open_for_install'
> 	from ./tool/rbinstall.rb:845:in `block in install_default_gem'
> 	from ./tool/rbinstall.rb:835:in `each'
> 	from ./tool/rbinstall.rb:835:in `install_default_gem'
> 	from ./tool/rbinstall.rb:799:in `block in <main>'
> 	from ./tool/rbinstall.rb:950:in `block in <main>'
> 	from ./tool/rbinstall.rb:947:in `each'
> 	from ./tool/rbinstall.rb:947:in `<main>'
> make: *** [do-install-all] Error 1
> 

and when i try 
`frum install 2.7.6 --with-openssl-dir=`brew --prefix openssl@1.1``

> ==> Downloading https://cache.ruby-lang.org/pub/ruby/2.7/ruby-2.7.6.tar.xz
==> Extracting ruby-2.7.6.tar.xz
==> Building Ruby 2.7.6
error: Can't build Ruby: make failed: . ./vm_opts.h
In file included from cont.c:16:
./mjit.h:131:17: warning: cast to smaller integer type 'enum rb_mjit_iseq_func' from 'mjit_func_t' (aka 'unsigned long (*)(struct rb_execution_context_struct *, struct rb_control_frame_struct *)') [-Wpointer-to-enum-cast]
        switch ((enum rb_mjit_iseq_func)func) {
                ^~~~~~~~~~~~~~~~~~~~~~~~~~~~
1 warning generated.
In file included from eval.c:20:
./mjit.h:131:17: warning: cast to smaller integer type 'enum rb_mjit_iseq_func' from 'mjit_func_t' (aka 'unsigned long (*)(struct rb_execution_context_struct *, struct rb_control_frame_struct *)') [-Wpointer-to-enum-cast]
        switch ((enum rb_mjit_iseq_func)func) {
                ^~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from gc.c:41:
./mjit.h:131:17: warning: cast to smaller integer type 'enum rb_mjit_iseq_func' from 'mjit_func_t' (aka 'unsigned long (*)(struct rb_execution_context_struct *, struct rb_control_frame_struct *)') [-Wpointer-to-enum-cast]
        switch ((enum rb_mjit_iseq_func)func) {
                ^~~~~~~~~~~~~~~~~~~~~~~~~~~~
1 warning generated.
In file included from iseq.c:30:
./mjit.h:131:17: warning: cast to smaller integer type 'enum rb_mjit_iseq_func' from 'mjit_func_t' (aka 'unsigned long (*)(struct rb_execution_context_struct *, struct rb_control_frame_struct *)') [-Wpointer-to-enum-cast]
        switch ((enum rb_mjit_iseq_func)func) {
                ^~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from mjit_compile.c:19:
./mjit.h:131:17: warning: cast to smaller integer type 'enum rb_mjit_iseq_func' from 'mjit_func_t' (aka 'unsigned long (*)(struct rb_execution_context_struct *, struct rb_control_frame_struct *)') [-Wpointer-to-enum-cast]
        switch ((enum rb_mjit_iseq_func)func) {
                ^~~~~~~~~~~~~~~~~~~~~~~~~~~~
1 warning generated.
1 warning generated.
1 warning generated.
In file included from ruby.c:55:
./mjit.h:131:17: warning: cast to smaller integer type 'enum rb_mjit_iseq_func' from 'mjit_func_t' (aka 'unsigned long (*)(struct rb_execution_context_struct *, struct rb_control_frame_struct *)') [-Wpointer-to-enum-cast]
        switch ((enum rb_mjit_iseq_func)func) {
                ^~~~~~~~~~~~~~~~~~~~~~~~~~~~
1 warning generated.
In file included from thread.c:77:
./mjit.h:131:17: warning: cast to smaller integer type 'enum rb_mjit_iseq_func' from 'mjit_func_t' (aka 'unsigned long (*)(struct rb_execution_context_struct *, struct rb_control_frame_struct *)') [-Wpointer-to-enum-cast]
        switch ((enum rb_mjit_iseq_func)func) {
                ^~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from vm.c:341:
./mjit.h:131:17: warning: cast to smaller integer type 'enum rb_mjit_iseq_func' from 'mjit_func_t' (aka 'unsigned long (*)(struct rb_execution_context_struct *, struct rb_control_frame_struct *)') [-Wpointer-to-enum-cast]
        switch ((enum rb_mjit_iseq_func)func) {
                ^~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from vm.c:350:
./vm_method.c:291:32: warning: cast to smaller integer type 'enum method_optimized_type' from 'void *' [-Wvoid-pointer-to-enum-cast]
            def->body.optimize_type = (enum method_optimized_type)opts;
                                      ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from vm_trace.c:28:
./mjit.h:131:17: warning: cast to smaller integer type 'enum rb_mjit_iseq_func' from 'mjit_func_t' (aka 'unsigned long (*)(struct rb_execution_context_struct *, struct rb_control_frame_struct *)') [-Wpointer-to-enum-cast]
        switch ((enum rb_mjit_iseq_func)func) {
                ^~~~~~~~~~~~~~~~~~~~~~~~~~~~
1 warning generated.
array.c:290:17: warning: expression result unused [-Wunused-value]
                RB_OBJ_WRITE(buff_owner_ary, &ptr[i+beg], argv[i]);
                ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
./internal.h:2711:34: note: expanded from macro 'RB_OBJ_WRITE'
#define RB_OBJ_WRITE(a, slot, b) UNALIGNED_MEMBER_ACCESS(rb_obj_write((VALUE)(a), (VALUE *)(slot), (VALUE)(b), __FILE__, __LINE__))
                                 ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
./internal.h:2703:5: note: expanded from macro 'UNALIGNED_MEMBER_ACCESS'
    unaligned_member_access_result; \
    ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
./include/ruby/ruby.h:1087:5: note: expanded from macro 'RARRAY_PTR_USE_TRANSIENT'
    expr; \
    ^~~~
1 warning generated.
In file included from mjit.c:18:
In file included from ./mjit_worker.c:76:
./mjit.h:131:17: warning: cast to smaller integer type 'enum rb_mjit_iseq_func' from 'mjit_func_t' (aka 'unsigned long (*)(struct rb_execution_context_struct *, struct rb_control_frame_struct *)') [-Wpointer-to-enum-cast]
        switch ((enum rb_mjit_iseq_func)func) {
                ^~~~~~~~~~~~~~~~~~~~~~~~~~~~
array.c:1200:2: warning: expression result unused [-Wunused-value]
        RB_OBJ_WRITE(target_ary, &ptr[idx], item);
        ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
./internal.h:2711:34: note: expanded from macro 'RB_OBJ_WRITE'
#define RB_OBJ_WRITE(a, slot, b) UNALIGNED_MEMBER_ACCESS(rb_obj_write((VALUE)(a), (VALUE *)(slot), (VALUE)(b), __FILE__, __LINE__))
                                 ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
./internal.h:2703:5: note: expanded from macro 'UNALIGNED_MEMBER_ACCESS'
    unaligned_member_access_result; \
    ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
./include/ruby/ruby.h:1087:5: note: expanded from macro 'RARRAY_PTR_USE_TRANSIENT'
    expr; \
    ^~~~
1 warning generated.
In file included from version.c:15:
./mjit.h:131:17: warning: cast to smaller integer type 'enum rb_mjit_iseq_func' from 'mjit_func_t' (aka 'unsigned long (*)(struct rb_execution_context_struct *, struct rb_control_frame_struct *)') [-Wpointer-to-enum-cast]
        switch ((enum rb_mjit_iseq_func)func) {
                ^~~~~~~~~~~~~~~~~~~~~~~~~~~~
1 warning generated.
2 warnings generated.
2 warnings generated.
ln -sf ../../../.ext/include/x86_64-darwin22/rb_mjit_min_header-2.7.6.h include/ruby-2.7.0/x86_64-darwin22/rb_mjit_min_header-2.7.6.h
Undefined symbols for architecture x86_64:
  "_rb_declare_transcoder", referenced from:
      _Init_transdb in transdb.o
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
make[1]: *** [.ext/x86_64-darwin22/enc/trans/transdb.bundle] Error 1
make[1]: *** Waiting for unfinished jobs....
make: *** [trans] Error 2
make: *** Waiting for unfinished jobs....
Undefined symbols for architecture x86_64:
  "_rb_enc_set_base", referenced from:
      _Init_encdb in encdb.o
  "_rb_enc_set_dummy", referenced from:
      _Init_encdb in encdb.o
  "_rb_encdb_alias", referenced from:
      _Init_encdb in encdb.o
  "_rb_encdb_declare", referenced from:
      _Init_encdb in encdb.o
  "_rb_encdb_dummy", referenced from:
      _Init_encdb in encdb.o
  "_rb_encdb_replicate", referenced from:
      _Init_encdb in encdb.o
  "_rb_encdb_set_unicode", referenced from:
      _Init_encdb in encdb.o
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
make[1]: *** [.ext/x86_64-darwin22/enc/encdb.bundle] Error 1
make[1]: *** Waiting for unfinished jobs....
make: *** [enc] Error 2
libffi_version: 3.4.0

this error came out. can you help me out please? :(

**Comment by aubri61** [comment]
`CFLAGS="-Wno-error=implicit-function-declaration" RUBY_CONFIGURE_OPTS='--with-readline-dir=/usr/local/opt/readline/' arch -x86_64 rbenv install 2.7.6`


this solved my problem

**Reply by aubri61** [reply]
this seems to install intel -x86 version ruby..
any other solutions?

**Reply by sanjayiothakur** [reply]
This was the only solution that worked for me. I'm on Ventura M2 Pro.

**Comment by mislav** [comment]
Sorry you're having trouble. Could you try installing 2.7.7 instead? It should be functionally the same as any other 2.7.x version, but 2.7.7 has some build fixes compared to 2.7.6.

```sh
RUBY_CONFIGURE_OPTS="--with-openssl-dir=$(brew --prefix openssl@1.1)" rbenv install 2.7.7
```

**Reply by SanjanaTailor** [reply]
@mislav - macOS monterary 12.6.3 

Build is getting failed.
ruby-build: using readline from homebrew
ruby-build: using gmp from homebrew

BUILD FAILED (macOS 12.6.3 using ruby-build 20230202)

Inspect or clean up the working tree at /var/folders/j0/fpcqrdsj21jf1x6nq957082c0000gq/T/ruby-build.20230207152106.21879.3sAvQJ
Results logged to /var/folders/j0/fpcqrdsj21jf1x6nq957082c0000gq/T/ruby-build.20230207152106.21879.log

Last 10 log lines:
linking miniruby
rb_mjit_header.h updated
generating encdb.h
make: ./miniruby: Permission denied
make: ./miniruby: Permission denied
make: *** [builtin_binary.inc] Error 1
make: *** Waiting for unfinished jobs....
make: *** [encdb.h] Error 1
./tool/ifchange "--timestamp=.rbconfig.time" rbconfig.rb rbconfig.tmp
rbconfig.rb updated

any help to resolve this issue please, inform.

**Reply by mislav** [reply]
@SanjanaTailor Your error `Permission denied` sounds very different than the original problem in this thread. Many people over the years have [experienced your problem](https://github.com/rbenv/ruby-build/discussions/1949), but I do not know the root cause.

You can try `MAKE_OPTS='-j1' rbenv install ...` (or `ruby-build …` if you are invoking it directly). The compile process will be slower but it might help you get around the problem.

**Comment by Bardiamist** [comment]
I also want  `rbenv install 2.7.6` but getting:

```
ruby-build: using readline from homebrew
ruby-build: using gmp from homebrew

BUILD FAILED (macOS 13.2.1 using ruby-build 20230208.1)

Inspect or clean up the working tree at /var/folders/vd/06k_fw_14vq2v3bkbjp6wjj80000gn/T/ruby-build.20230220182332.41440.bDdsLn
Results logged to /var/folders/vd/06k_fw_14vq2v3bkbjp6wjj80000gn/T/ruby-build.20230220182332.41440.log

Last 10 log lines:
compiling vm_trace.c
make: /opt/homebrew/bin/gmkdir: Permission denied
make: *** [.ext/.timestamp/.enc-trans.time] Error 1
make: *** Waiting for unfinished jobs....
In file included from vm_trace.c:28:
./mjit.h:131:17: warning: cast to smaller integer type 'enum rb_mjit_iseq_func' from 'mjit_func_t' (aka 'unsigned long (*)(struct rb_execution_context_struct *, struct rb_control_frame_struct *)') [-Wpointer-to-enum-cast]
        switch ((enum rb_mjit_iseq_func)func) {
                ^~~~~~~~~~~~~~~~~~~~~~~~~~~~
1 warning generated.
2 warnings generated.
```

**Comment by RayJason** [comment]
My mac is M1 Pro, Ventura 13.2.1.
Unfortunately, the answer in the comment didn't work for me. I solved this problem by rvm. I had to uninstall rbenv and ruby-build.

```bash
$ \curl -sSL https://get.rvm.io | bash -s stable # install rvm

$ CFLAGS="-Wno-error=implicit-function-declaration" rvm install 2.7.2
```

**Comment by prateeksen** [comment]
Passing RUBY_CFLAGS="-w" worked for me on MacOS Ventura M1 Max. 
example: 
`RUBY_CFLAGS="-w" rbenv install 2.4.0`

**Reply by aleena-adnan1** [reply]
but where to pass it ?

**Reply by aleena-adnan1** [reply]
i tried it but no luck :(

**Comment by Bardiamist** [comment]
Tried `RUBY_CFLAGS="-w" rbenv install 2.7.6` but got:
```
BUILD FAILED (macOS 13.2.1 using ruby-build 20230222)

Inspect or clean up the working tree at /var/folders/vd/06k_fw_14vq2v3bkbjp6wjj80000gn/T/ruby-build.20230224114402.70244.qwOccC
Results logged to /var/folders/vd/06k_fw_14vq2v3bkbjp6wjj80000gn/T/ruby-build.20230224114402.70244.log

Last 10 log lines:
compiling dmyenc.c
compiling dln.c
compiling localeinit.c
make: /opt/homebrew/bin/gmkdir: Permission denied
make: /opt/homebrew/bin/gmkdir: Permission denied
make: *** [.ext/.timestamp/.RUBYCOMMONDIR.time] Error 1
make: *** Waiting for unfinished jobs....
make: *** [.ext/.timestamp/arm64-darwin22/.time] Error 1
compiling array.c
config.status: creating ruby-runner.h
```

**Reply by prateeksen** [reply]
Trying cleaning up the working tree as mentions and check permissions for /opt/homebrew/

**Reply by Bardiamist** [reply]
> check permissions for /opt/homebrew/

What and how need to check?
`sudo chown -R $(whoami) $(brew --prefix)/*` and `sudo chown -R "$USER":admin /opt` and `sudo chown -R "$USER":admin /opt/homebrew/` not helped

**Reply by Bardiamist** [reply]
Still waiting for help
M1 Max
Ventura 13.2.1

**Reply by prateeksen** [reply]
@Bardiamist Please check whether you have rosetta installed on your mac.
Probably that is required, I have that running on my system.
Run command : `/usr/bin/pgrep -q oahd && echo Yes || echo No`
Ref: https://apple.stackexchange.com/questions/427970/how-to-tell-if-m1-mac-has-rosetta-installed

**Reply by Bardiamist** [reply]
@prateek-ap It returns "Yes"

**Reply by prateeksen** [reply]
Hmm.. That's quite weird.. You might need to dig more into it then, I might not be able to help anymore from here.

**Reply by krzysztofkeczkowski** [reply]
`RUBY_CONFIGURE_OPTS="--with-openssl-dir=$(brew --prefix openssl@1.1)" rbenv install 3.2.0` worked for me.

M1 macbook air
Ventura 13.2.1

**Reply by rmtolmach** [reply]
> Inspect or clean up the working tree at /var/folders/vd/06k_fw_14vq2v3bkbjp6wjj80000gn/T/ruby-build.20230224114402.70244.qwOccC

`cd` into the file specified in the output. In this example, `/var/folders/vd/06k_fw_14vq2v3bkbjp6wjj80000gn/T/ruby-build.20230224114402.70244.qwOccC`. Then `cd` into the ruby directory in _that_ folder. (`ruby-2.7.6` in this example.) Run `make install`. Make sure there aren't any failures in the output (in my specific case, there was an `openssl` error and brew installing the version I needed fixed it). Restart your terminal.

**Comment by krzysztofkeczkowski** [comment]
`RUBY_CONFIGURE_OPTS="--with-openssl-dir=$(brew --prefix openssl@1.1)" rbenv install 3.2.0` worked for me.

M1 macbook air
Ventura 13.2.1

**Comment by alexsorokoletov** [comment, marked as answer]
Here's what worked for me on Ventura M2 for `rbenv` + `ruby 2.7.5`

```
brew unlink openssl@3  
brew install openssl@1.1  

echo 'export PATH="/usr/local/opt/openssl@1.1/bin:$PATH"' >> ~/.zshrc   
export LDFLAGS="-L/usr/local/opt/openssl@1.1/lib"                                                     
export CPPFLAGS="-I/usr/local/opt/openssl@1.1/include"
export PKG_CONFIG_PATH="/usr/local/opt/openssl@1.1/lib/pkgconfig"
RUBY_CONFIGURE_OPTS="--with-openssl-dir=$(brew --prefix openssl@1.1)" arch -x86_64 rbenv install 2.7.5
```

**Reply by mislav** [reply]
Thanks for sharing, but permanently editing one's PATH and temporarily setting LDFLAGS, CPPFLAGS, ~and PKG_CONFIG_PATH~ are not necessary if `--with-openssl-dir` was used. Furthermore, your solution does not point out that even though you have the M2 chip, you installed Ruby for amd64 architecture, which is not optimal because it does not match the native architecture. However, if that is the only approach that worked for you, then I guess that's fair. Finally, `brew unlink openssl@3` should not be necessary either because Homebrew's openssl is "keg-only", which means it's never linked into PATH by default to avoid overriding system OpenSSL.

Please note that Ruby 2.7 is "End-of-life" and that it will get progressively harder and harder to install on newer computers and versions of macOS. It should be way easier to upgrade old applications than to reliably install old versions of Ruby.

**Reply by alexsorokoletov** [reply]
Totally agree with everything @mislav said!
I should've included the disclaimer (thank you @mislav) that this is more a "desperate workaround" rather than a valid path forward.

For my usecase, I needed an exact version of ruby running on a dev machine so that we can verify the baseline of the rails app (aka run tests, etc etc) and then migrate to newer rails/ruby knowing the baseline.
Same with the performance.

Thank you for adding this note and highlighting important aspects of this!

**Reply by MayneX** [reply]
Tried with 2.6.7 and Intel. Failed:
```
vm.c:2489:34: warning: expression does not compute the number of elements in this array; element type is 'const int', not 'VALUE' (aka 'unsigned long') [-Wsizeof-array-div]
                             sizeof(ec->machine.regs) / sizeof(VALUE));
                                    ~~~~~~~~~~~~~~~~  ^
vm.c:2489:34: note: place parentheses around the 'sizeof(VALUE)' expression to silence this warning
61 warnings and 1 error generated.
make: *** [vm.o] Error 1
make: *** Waiting for unfinished jobs....
32 warnings generated.
2 warnings generated.
```

It works after adding `export CFLAGS="-Wno-error=implicit-function-declaration"`

```
brew unlink openssl@3  
brew install openssl@1.1  

echo 'export PATH="/usr/local/opt/openssl@1.1/bin:$PATH"' >> ~/.zshrc
export LDFLAGS="-L/usr/local/opt/openssl@1.1/lib"            export CPPFLAGS="-I/usr/local/opt/openssl@1.1/include"
export CFLAGS="-Wno-error=implicit-function-declaration"
export PKG_CONFIG_PATH="/usr/local/opt/openssl@1.1/lib/pkgconfig"
RUBY_CONFIGURE_OPTS="--with-openssl-dir=$(brew --prefix openssl@1.1)" arch -x86_64 rbenv install 2.6.7
```

**Reply by Jonathan0wh** [reply]
Please change `/usr/local/opt/openssl@1.1/` to your openssl@1.1 path like `/opt/homebrew/opt/openssl@1.1` if you installed openssl@1.1 using brew, or you'll get permission denied error

**Reply by mislav** [reply]
@MayneX: Again, per my above comment: you do not need to set PATH, LDFLAGS, CPPFLAGS, ~nor PKG_CONFIG_PATH~ to link to Homebrew's OpenSSL. Just setting `RUBY_CONFIGURE_OPTS="--with-openssl-dir=..."` will suffice.

Furthermore, if you are really on an Intel chip, then your system is already `x86_64` architecture. There is no need to ever do `arch -x86_64`.

Plus, as @Jonathan0wh points out, it's better for maximum portability to always use the result of `$(brew --prefix openssl@1.1)` instead of hardcoding paths like `/usr/local/opt/openssl@1.1` in your scripts.

**Comment by tomwang57** [comment]
I do config openssl@1.1 and still failed with:
```
ossl_ssl.c:1880:52ossl_pkey_rsa.c:942:5:: error: use of undeclared identifier 'RSA_SSLV23_PADDING'
    DefRSAConst(SSLV23_PADDING);
```

config info:
```
ruby-build: using readline from homebrew
ruby-build: using libyaml from homebrew
ruby-build: using gmp from homebrew
-> ./configure "--prefix=$HOME/.rbenv/versions/2.7.4" --enable-shared --with-readline-dir=/opt/homebrew/opt/readline --with-libyaml-dir=/opt/homebrew/opt/libyaml --with-gmp-dir=/opt/homebrew/opt/gmp --with-openssl-dir=/opt/homebrew/opt/openssl@1.1
-> make -j 8

BUILD FAILED (macOS 13.4 on arm64 using ruby-build 20231107)
```

**Reply by mislav** [reply]
Do you by chance also have `openssl@3` installed with Homebrew? You can try temporarily `brew unlink openssl@3` and try again. I think that Ruby 2.7 might be trying to pick up a newer openssl version present on the system instead of the one passed explicitly with `--with-openssl-dir=/opt/homebrew/opt/openssl@1.1`. But, that's just guessing on my part.

Ref. https://github.com/rbenv/ruby-build/issues/2300

**Comment by tomwang57** [comment]
**Ultimate solution:** 
https://github.com/postmodern/ruby-install/issues/409
<img width="870" alt="Pasted Graphic" src="https://github.com/rbenv/ruby-build/assets/144871969/0d6700dc-2812-416d-a2a9-7f58f5ea7745">
